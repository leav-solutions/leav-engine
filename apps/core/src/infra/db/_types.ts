// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
