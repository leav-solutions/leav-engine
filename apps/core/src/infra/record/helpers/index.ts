// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as filterTypes} from './filterTypes';
export {default as getClassifyingFiltersVariableQueryPart} from './getClassifyingFiltersVariableQueryPart';
export {default as getSearchVariableName} from './getSearchVariableName';
export {default as getSearchVariablesQueryPart} from './getSearchVariablesQueryPart';
