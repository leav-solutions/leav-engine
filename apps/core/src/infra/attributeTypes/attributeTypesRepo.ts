// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AqlLiteral, AqlQuery, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeTypes, IAttribute} from '../../_types/attribute';
import {AttributeCondition, IRecordFilterOption, IRecordSort} from '../../_types/record';
import {IValue, IValuesOptions} from '../../_types/value';

// To avoid some cyclic dependencies issues, we have to pass repo along attribute props
export interface IAttributeWithRepo extends IAttribute {
    _repo: IAttributeTypeRepo;
}

export interface IAttributeTypesRepo {
    getTypeRepo?(attribute: IAttribute): IAttributeTypeRepo;
}

export type GetConditionPartParentFunc = (
    condition: AttributeCondition,
    value: string | number | boolean,
    attribute: IAttribute
) => GetConditionPartFunc;

export type GetConditionPartFunc = (valueIdentifier: string | AqlLiteral) => GeneratedAqlQuery;
/**
 * Define interface used for all attribute type specific files
 */
export interface IAttributeTypeRepo {
    createValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Update an existing value. Field "id" is expected on the value
     */
    updateValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Delete an existing value. Field "id" is expected on the value
     */
    deleteValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Get all values for given record and attribute
     *
     * @return Array<{}>    Return an empty array if no value found
     */
    getValues({
        library,
        recordId,
        attribute,
        forceGetAllValues,
        options,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        forceGetAllValues?: boolean;
        options?: IValuesOptions;
        ctx: IQueryInfos;
    }): Promise<IValue[]>;

    /**
     * Return a specific value based on its ID. Field "id" is expect on the value
     *
     * @return {}   Return null if no value found
     */
    getValueById({
        library,
        recordId,
        attribute,
        valueId,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        valueId: string;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Return AQL query part to filter on this attribute. If will be concatenate with other filters and full query
     */
    filterQueryPart(attributes: IAttributeWithRepo[], filter: IRecordFilterOption, parentIdentifier?: string): AqlQuery;

    /**
     * Return AQL query part to sort on this attribute
     */
    sortQueryPart({attributes, order}: IRecordSort): AqlQuery;

    /**
     * Clear all values of given attribute. Can be used to cleanup values when an attribute is deleted for example.
     *
     * @param attribute
     * @return Promise<number> TRUE if operation succeed
     */
    clearAllValues({attribute, ctx}: {attribute: IAttribute; ctx: IQueryInfos}): Promise<boolean>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';
export const BASE_QUERY_IDENTIFIER = 'r';

interface IDeps {
    'core.infra.attributeTypes.attributeSimple'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeSimpleLink'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeAdvanced'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeAdvancedLink'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeTree'?: IAttributeTypeRepo;
}

export default function ({
    'core.infra.attributeTypes.attributeSimple': attributeSimpleRepo = null,
    'core.infra.attributeTypes.attributeSimpleLink': attributeSimpleLinkRepo = null,
    'core.infra.attributeTypes.attributeAdvanced': attributeAdvancedRepo = null,
    'core.infra.attributeTypes.attributeAdvancedLink': attributeAdvancedLinkRepo = null,
    'core.infra.attributeTypes.attributeTree': attributeTreeRepo = null
}: IDeps = {}): IAttributeTypesRepo {
    return {
        getTypeRepo(attribute) {
            let attrTypeRepo: IAttributeTypeRepo;
            switch (attribute.type) {
                case AttributeTypes.SIMPLE:
                    attrTypeRepo = attributeSimpleRepo;
                    break;
                case AttributeTypes.SIMPLE_LINK:
                    attrTypeRepo = attributeSimpleLinkRepo;
                    break;
                case AttributeTypes.ADVANCED:
                    attrTypeRepo = attributeAdvancedRepo;
                    break;
                case AttributeTypes.ADVANCED_LINK:
                    attrTypeRepo = attributeAdvancedLinkRepo;
                    break;
                case AttributeTypes.TREE:
                    attrTypeRepo = attributeTreeRepo;
                    break;
            }

            return attrTypeRepo;
        }
    };
}
