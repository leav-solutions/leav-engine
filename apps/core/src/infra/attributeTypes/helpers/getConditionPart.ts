// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, AqlLiteral, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import moment from 'moment';
import {AttributeFormats, IAttribute} from '../../../_types/attribute';
import {AttributeCondition, IDateFilterValue} from '../../../_types/record';

export type GetConditionPart = (
    valueIdentifier: string | AqlLiteral,
    condition: AttributeCondition,
    value: string | number | boolean,
    attribute: IAttribute
) => GeneratedAqlQuery;

export default function (): GetConditionPart {
    return (
        valueIdentifier: string | AqlLiteral,
        condition: AttributeCondition,
        value: string | number | boolean | IDateFilterValue,
        attribute: IAttribute
    ): GeneratedAqlQuery => {
        const valueField = typeof valueIdentifier === 'string' ? aql.literal(valueIdentifier) : valueIdentifier;

        switch (condition) {
            case AttributeCondition.EQUAL: {
                const cond =
                    attribute.format === AttributeFormats.DATE
                        ? aql`DATE_COMPARE(${valueField} * 1000, ${Number(value) * 1000}, "years", "days") == true`
                        : aql`${valueField} == ${value}`;

                return cond;
            }
            case AttributeCondition.NOT_EQUAL:
                return attribute.format === AttributeFormats.DATE
                    ? aql`DATE_COMPARE(${valueField} * 1000, ${Number(value) * 1000}, "years", "days") == false`
                    : aql`${valueField} != ${value}`;
            case AttributeCondition.BEGIN_WITH:
                return aql`${valueField} LIKE ${`${value}%`}`;
            case AttributeCondition.END_WITH:
                return aql`${valueField} LIKE ${`%${value}`}`;
            case AttributeCondition.CONTAINS: {
                return attribute.format === AttributeFormats.DATE_RANGE
                    ? aql`(${Number(value)} >= ${valueField}.from AND ${Number(value)} <= ${valueField}.to)`
                    : aql`${valueField} LIKE ${`%${value}%`}`;
            }
            case AttributeCondition.NOT_CONTAINS:
                return aql`${valueField} NOT LIKE ${`%${value}%`}`;
            case AttributeCondition.GREATER_THAN:
                return aql`${valueField} > ${Number(value)}`;
            case AttributeCondition.LESS_THAN:
                return aql`(${valueField} != null AND ${valueField} < ${Number(value)})`;
            case AttributeCondition.IS_EMPTY:
                return aql`${valueField} == null`;
            case AttributeCondition.IS_NOT_EMPTY:
                return aql`${valueField} != null`;
            case AttributeCondition.BETWEEN:
                return aql`(
                            ${valueField} >= ${Number((value as IDateFilterValue).from)}
                            AND ${valueField} <= ${Number((value as IDateFilterValue).to)}
                        )`;
            case AttributeCondition.TODAY:
                return aql`DATE_COMPARE(${valueField} * 1000, DATE_NOW(), "years", "days") == true`;
            case AttributeCondition.YESTERDAY:
                return aql`DATE_COMPARE(
                            ${valueField} * 1000,
                            DATE_SUBTRACT(DATE_NOW(), 1, "day"),
                            "years",
                            "days"
                        ) == true`;
            case AttributeCondition.TOMORROW:
                return aql`DATE_COMPARE(
                            ${valueField} * 1000,
                            DATE_ADD(DATE_NOW(), 1, "day"),
                            "years",
                            "days"
                        ) == true`;
            case AttributeCondition.NEXT_MONTH: {
                const now = moment().unix();
                const nextMonth = moment().add(31, 'days').unix();
                return aql`(${valueField} >= ${now} AND ${valueField} <=${nextMonth})`;
            }
            case AttributeCondition.LAST_MONTH: {
                const now = moment().unix();
                const lastMonth = moment().subtract(31, 'days').unix();
                return aql`(${valueField} >= ${lastMonth} AND ${valueField} <=${now})`;
            }
            case AttributeCondition.START_ON:
                return aql`DATE_COMPARE(${valueField}.from * 1000, ${Number(value) * 1000}, "years", "days") == true`;
            case AttributeCondition.START_BEFORE:
                return aql`(${valueField}.from != null AND ${valueField}.from < ${Number(value)})`;
            case AttributeCondition.START_AFTER:
                return aql`${valueField}.from > ${Number(value)}`;
            case AttributeCondition.END_ON:
                return aql`DATE_COMPARE(${valueField}.to * 1000, ${Number(value) * 1000}, "years", "days") == true`;
            case AttributeCondition.END_BEFORE:
                return aql`(${valueField}.to != null AND ${valueField}.to < ${Number(value)})`;
            case AttributeCondition.END_AFTER:
                return aql`${valueField}.to > ${Number(value)}`;
            default:
                return aql`${valueField} == ${value}`;
        }
    };
}
