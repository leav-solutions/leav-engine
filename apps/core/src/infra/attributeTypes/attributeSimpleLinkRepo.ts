// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {AttributeFormats, IAttribute} from '../../_types/attribute';
import {AttributeCondition, IRecordFilterOption, IRecordSort} from '../../_types/record';
import {ILinkValue, IStandardValue} from '../../_types/value';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {BASE_QUERY_IDENTIFIER, IAttributeTypeRepo, IAttributeWithRepo} from './attributeTypesRepo';
import {GetConditionPart} from './helpers/getConditionPart';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.infra.attributeTypes.attributeSimple'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.helpers.getConditionPart'?: GetConditionPart;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.infra.attributeTypes.attributeSimple': attributeSimpleRepo = null,
    'core.infra.attributeTypes.helpers.getConditionPart': getConditionPart = null
}: IDeps = {}): IAttributeTypeRepo {
    function _getExtendedFilterPart(attributes: IAttribute[], linkedValue: AqlQuery): AqlQuery {
        return attributes
            .map(a => a.id)
            .slice(2)
            .reduce((acc, value, i) => {
                acc.push(aql`TRANSLATE(${value}, ${i ? acc[acc.length - 1] : aql`${linkedValue}`})`);
                if (i) {
                    acc.shift();
                }
                return acc;
            }, [])[0];
    }

    const _buildLinkValue = (savedValue: IStandardValue, attribute: IAttribute): ILinkValue => ({
        ...savedValue,
        value: {id: savedValue.value, library: attribute.linked_library}
    });

    return {
        async createValue(args): Promise<ILinkValue> {
            const savedVal = await attributeSimpleRepo.createValue(args);

            return _buildLinkValue(savedVal, args.attribute);
        },
        async updateValue(args): Promise<ILinkValue> {
            const savedVal = await attributeSimpleRepo.updateValue(args);

            return _buildLinkValue(savedVal, args.attribute);
        },
        async deleteValue(args): Promise<ILinkValue> {
            const deletedValue = await attributeSimpleRepo.deleteValue(args);

            return _buildLinkValue(deletedValue, args.attribute);
        },
        // To get values from advanced reverse link attribute into simple link.
        async getReverseValues({advancedLinkAttr, value, forceGetAllValues = false, ctx}): Promise<ILinkValue[]> {
            const libCollec = dbService.db.collection(advancedLinkAttr.linked_library);
            const queryParts = [];

            queryParts.push(aql`
                FOR r IN ${libCollec}
                    FILTER r.${(advancedLinkAttr.reverse_link as IAttribute)?.id} == ${value}`);

            const limitOne = aql.literal(!advancedLinkAttr.multiple_values && !forceGetAllValues ? 'LIMIT 1' : '');

            queryParts.push(aql`
                ${limitOne}
                RETURN r
            `);

            const query = aql.join(queryParts);
            const res = await dbService.execute({query, ctx});

            return res.map(r => ({id_value: null, value: dbUtils.cleanup(r), created_by: null, modified_by: null}));
        },
        async getValues({library, recordId, attribute, ctx}): Promise<ILinkValue[]> {
            const libCollec = dbService.db.collection(library);
            const linkedLibCollec = dbService.db.collection(attribute.linked_library);

            const res = await dbService.execute({
                query: aql`
                    FOR r IN ${libCollec}
                        FILTER r._key == ${recordId}
                        FOR l IN ${linkedLibCollec}
                            FILTER r.${attribute.id} == l._key
                            RETURN l
                `,
                ctx
            });

            return res
                .slice(0, 1)
                .map(r => ({id_value: null, value: dbUtils.cleanup(r), created_by: null, modified_by: null}));
        },
        sortQueryPart({attributes, order}: {attributes: IAttribute[]; order: string}): AqlQuery {
            const linkedLibCollec = dbService.db.collection(attributes[0].linked_library);
            const linked = !attributes[1]
                ? {id: '_key', format: AttributeFormats.TEXT}
                : attributes[1].id === 'id'
                ? {...attributes[1], id: '_key'}
                : attributes[1];

            const linkedValue = aql`
                FIRST(FOR l IN ${linkedLibCollec}
                    FILTER TO_STRING(r.${attributes[0].id}) == l._key
                RETURN l.${linked.id})
            `;

            const query: AqlQuery =
                linked.format !== AttributeFormats.EXTENDED
                    ? aql`SORT ${linkedValue} ${order}`
                    : aql`SORT ${_getExtendedFilterPart(attributes, linkedValue)} ${order}`;

            return query;
        },
        filterQueryPart(
            attributes: IAttributeWithRepo[],
            filter: IRecordFilterOption,
            parentIdentifier = BASE_QUERY_IDENTIFIER
        ): AqlQuery {
            const linkedLibCollec = dbService.db.collection(attributes[0].linked_library);
            const linked = !attributes[1]
                ? {id: '_key', format: AttributeFormats.TEXT}
                : attributes[1].id === 'id'
                ? {...attributes[1], id: '_key'}
                : attributes[1];

            const retrieveValue = aql`FOR l IN ${linkedLibCollec}
                FILTER TO_STRING(r.${attributes[0].id}) == l._key`;
            const returnValue = aql`RETURN l`;

            let query: AqlQuery;
            const linkValIdentifier = aql.literal(`${parentIdentifier}linkVal`);
            if (
                [
                    AttributeCondition.VALUES_COUNT_EQUAL,
                    AttributeCondition.VALUES_COUNT_GREATER_THAN,
                    AttributeCondition.VALUES_COUNT_LOWER_THAN
                ].includes(filter.condition as AttributeCondition)
            ) {
                let conditionApplied = filter.condition;
                switch (filter.condition) {
                    case AttributeCondition.VALUES_COUNT_EQUAL:
                        conditionApplied = AttributeCondition.EQUAL;
                        break;
                    case AttributeCondition.VALUES_COUNT_GREATER_THAN:
                        conditionApplied = AttributeCondition.GREATER_THAN;
                        break;
                    case AttributeCondition.VALUES_COUNT_LOWER_THAN:
                        conditionApplied = AttributeCondition.LESS_THAN;
                        break;
                }

                query = aql.join([
                    aql`LET ${linkValIdentifier} = `,
                    aql`COUNT(`,
                    retrieveValue,
                    returnValue,
                    aql`) ? 1 : 0`,
                    aql`FILTER ${getConditionPart(
                        linkValIdentifier,
                        conditionApplied as AttributeCondition,
                        filter.value,
                        attributes[0]
                    )}`
                ]);
            } else if (
                filter.condition === AttributeCondition.IS_EMPTY ||
                filter.condition === AttributeCondition.IS_NOT_EMPTY
            ) {
                query = aql.join([
                    aql`LET ${linkValIdentifier} = `,
                    aql`FIRST(`,
                    retrieveValue,
                    returnValue,
                    aql`)`,
                    aql`FILTER ${getConditionPart(linkValIdentifier, filter.condition, filter.value, attributes[0])}`
                ]);
            } else {
                const filterLinkedValue = attributes[1]
                    ? attributes[1]._repo.filterQueryPart([...attributes].splice(1), filter, 'l')
                    : null;

                const linkedValue = aql.join([aql`FIRST(`, retrieveValue, filterLinkedValue, returnValue, aql`)`]);

                query =
                    linked.format !== AttributeFormats.EXTENDED
                        ? aql`FILTER ${linkedValue} != null`
                        : aql`FILTER ${_getExtendedFilterPart(attributes, linkedValue)} != null`;
            }

            return query;
        },
        async clearAllValues(args): Promise<boolean> {
            return true;
        }
    };
}
