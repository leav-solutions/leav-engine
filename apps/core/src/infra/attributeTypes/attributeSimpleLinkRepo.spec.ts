// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {aql} from 'arangojs/aql';
import {DocumentCollection} from 'arangojs/collection';
import {IDbUtils} from 'infra/db/dbUtils';
import {IFilterTypesHelper} from 'infra/record/helpers/filterTypes';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeTypes} from '../../_types/attribute';
import {AttributeCondition} from '../../_types/record';
import {mockAttrSimpleLink} from '../../__tests__/mocks/attribute';
import attributeSimpleLinkRepo from './attributeSimpleLinkRepo';
import {IAttributeTypeRepo} from './attributeTypesRepo';

describe('AttributeSimpleLinkRepo', () => {
    const mockAttribute = {
        id: 'test_simple_link_attr',
        type: AttributeTypes.SIMPLE_LINK,
        linked_library: 'test_linked_lib'
    };

    const mockAttrSimpleRepo: Mockify<IAttributeTypeRepo> = {
        createValue: null,
        updateValue: null,
        deleteValue: null,
        getValueById: null,
        getValues: null,
        sortQueryPart: null,
        clearAllValues: null
    };
    const ctx: IQueryInfos = {
        userId: '0',
        queryId: 'attributeSimpleLinkRepoTest'
    };

    describe('createValue', () => {
        test('Should create a simple link value', async function () {
            const updatedValueData = {
                value: '123456'
            };

            const attrSimpleRepo = {
                ...mockAttrSimpleRepo,
                createValue: global.__mockPromise(updatedValueData),
                updateValue: global.__mockPromise(updatedValueData)
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.attributeTypes.attributeSimple': attrSimpleRepo as IAttributeTypeRepo
            });

            const createdVal = await attrRepo.createValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 123456
                },
                ctx
            });

            expect(attrSimpleRepo.createValue.mock.calls.length).toBe(1);
            expect(attrSimpleRepo.createValue).toBeCalledWith({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 123456
                },
                ctx
            });

            expect(createdVal).toMatchObject({
                ...updatedValueData,
                value: {
                    id: '123456',
                    library: 'test_linked_lib'
                }
            });
        });
    });

    describe('deleteValue', () => {
        test('Should delete a value', async () => {
            const deletedValueData = {
                value: '123456'
            };

            const attrSimpleRepo = {
                ...mockAttrSimpleRepo,
                deleteValue: global.__mockPromise(deletedValueData)
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.attributeTypes.attributeSimple': attrSimpleRepo as IAttributeTypeRepo
            });

            const deletedVal = await attrRepo.deleteValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 123456
                },
                ctx
            });

            expect(attrSimpleRepo.deleteValue.mock.calls.length).toBe(1);
            expect(attrSimpleRepo.deleteValue).toBeCalledWith({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 123456
                },
                ctx
            });

            expect(deletedVal).toMatchObject({
                ...deletedValueData,
                value: {
                    id: '123456',
                    library: 'test_linked_lib'
                }
            });
        });
    });

    describe('getValues', () => {
        test('Should return values for simple link attribute', async function () {
            const queryRes = [
                {
                    _key: '987654',
                    _id: 'images/987654',
                    _rev: '_WgJhrXO--_',
                    created_at: 1521475225,
                    modified_at: 1521475225
                },
                {
                    _key: '987655',
                    _id: 'images/987655',
                    _rev: '_WgJhrXO--_',
                    created_at: 1521475225,
                    modified_at: 1521475225
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(queryRes)
            };

            const mockCleanupRes = jest.fn().mockReturnValue({
                id: 987654,
                created_at: 1521475225,
                modified_at: 1521475225
            });

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();
            expect(mockDbUtils.cleanup.mock.calls.length).toBe(1);

            expect(values.length).toBe(1);

            expect(values[0]).toMatchObject({
                id_value: null,
                value: {
                    id: 987654,
                    created_at: 1521475225,
                    modified_at: 1521475225
                }
            });
        });
    });

    describe('getReverseValues', () => {
        test('Should return values for advanced reverse link attribute into simple link', async function () {
            const queryRes = [
                {
                    _key: '987654',
                    _id: 'images/987654',
                    _rev: '_WgJhrXO--_',
                    created_at: 1521475225,
                    modified_at: 1521475225
                },
                {
                    _key: '987655',
                    _id: 'images/987655',
                    _rev: '_WgJhrXO--_',
                    created_at: 1521475225,
                    modified_at: 1521475225
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(queryRes)
            };

            const mockCleanupRes = jest.fn().mockReturnValue({
                id: 987654,
                created_at: 1521475225,
                modified_at: 1521475225
            });

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: mockCleanupRes
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();
            expect(mockDbUtils.cleanup.mock.calls.length).toBe(1);

            expect(values.length).toBe(1);

            expect(values[0]).toMatchObject({
                id_value: null,
                value: {
                    id: 987654,
                    created_at: 1521475225,
                    modified_at: 1521475225
                }
            });
        });
    });

    describe('filterValueQueryPart', () => {
        const mockFilterTypesHelper: Mockify<IFilterTypesHelper> = {
            isCountFilter: jest.fn().mockReturnValue(false)
        };

        test('Should return query to retrieve value to filter on', () => {
            const mockDb = {
                collection: jest.fn().mockReturnValue({} as DocumentCollection)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const mockRepo: Mockify<IAttributeTypeRepo> = {
                filterValueQueryPart: jest.fn().mockReturnValue(aql``)
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper,
                'core.infra.db.dbService': mockDbServ
            });

            const valueQuery = attrRepo.filterValueQueryPart(
                [
                    {id: 'label', type: AttributeTypes.SIMPLE_LINK, _repo: mockRepo as IAttributeTypeRepo},
                    {id: 'linked', type: AttributeTypes.SIMPLE, _repo: mockRepo as IAttributeTypeRepo}
                ],
                {condition: AttributeCondition.EQUAL, value: 'MyLabel'}
            );

            expect(valueQuery).toMatchSnapshot();
        });

        test('Should return query to retrieve value to filter on for reverse link', async () => {
            const mockDb = {
                collection: jest.fn().mockReturnValue({} as DocumentCollection)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const mockRepo: Mockify<IAttributeTypeRepo> = {
                filterValueQueryPart: jest.fn().mockReturnValue(aql`<VALUE QUERY PART>`)
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper,
                'core.infra.db.dbService': mockDbServ
            });

            const valueQuery = attrRepo.filterValueQueryPart(
                [
                    {
                        id: 'linked_from',
                        type: AttributeTypes.SIMPLE_LINK,
                        reverse_link: {...mockAttrSimpleLink},
                        _repo: mockRepo as IAttributeTypeRepo
                    },
                    {id: 'label', type: AttributeTypes.ADVANCED, _repo: mockRepo as IAttributeTypeRepo}
                ],
                {condition: AttributeCondition.EQUAL, value: 'MyLabel'}
            );

            expect(valueQuery).toMatchSnapshot();
        });

        test('Should return query to retrieve value to filter on for "count" filter', async () => {
            const mockDbServ = {
                db: new Database()
            };

            const mockFilterTypesHelperCount: Mockify<IFilterTypesHelper> = {
                isCountFilter: jest.fn().mockReturnValue(true)
            };

            const mockRepo: Mockify<IAttributeTypeRepo> = {
                filterValueQueryPart: jest.fn().mockReturnValue(aql`<VALUE QUERY PART>`)
            };

            const attrRepo = attributeSimpleLinkRepo({
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelperCount as IFilterTypesHelper,
                'core.infra.db.dbService': mockDbServ
            });

            const valueQuery = attrRepo.filterValueQueryPart(
                [
                    {
                        id: 'linked_from',
                        type: AttributeTypes.SIMPLE_LINK,
                        _repo: mockRepo as IAttributeTypeRepo
                    }
                ],
                {condition: AttributeCondition.VALUES_COUNT_EQUAL, value: '42'}
            );

            expect(valueQuery).toMatchSnapshot();
        });
    });

    describe('sortQueryPart', () => {
        test('Should return simple link sort', () => {
            const mockDb = {
                collection: jest.fn().mockReturnValue({} as DocumentCollection)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const attrRepo = attributeSimpleLinkRepo({'core.infra.db.dbService': mockDbServ});
            const filter = attrRepo.sortQueryPart({
                attributes: [
                    {id: 'label', type: AttributeTypes.SIMPLE_LINK},
                    {id: 'linked', type: AttributeTypes.SIMPLE}
                ],
                order: 'ASC'
            });

            expect(filter.query).toMatch(/^SORT/);
            expect(filter).toMatchSnapshot();
        });
    });
});
