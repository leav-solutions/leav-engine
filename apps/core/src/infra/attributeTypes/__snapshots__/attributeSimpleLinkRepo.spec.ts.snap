// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeSimpleLinkRepo filterValueQueryPart Should return query to retrieve value to filter on 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": {},
    "value1": "label",
  },
  "query": "FLATTEN( FOR l IN @value0
                    FILTER TO_STRING(r.@value1) == l._key 
                    LET rlinkVal = ()
                    RETURN rlinkVal
                 )",
}
`;

exports[`AttributeSimpleLinkRepo filterValueQueryPart Should return query to retrieve value to filter on for "count" filter 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": "linked_from",
  },
  "query": "COUNT(r.@value0) ? 1 : 0",
}
`;

exports[`AttributeSimpleLinkRepo filterValueQueryPart Should return query to retrieve value to filter on for reverse link 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": {},
    "value1": "linked_from",
  },
  "query": "FLATTEN( FOR l IN @value0
                    FILTER TO_STRING(r.@value1) == l._key 
                    LET rlinkVal = (<VALUE QUERY PART>)
                    RETURN rlinkVal
                 )",
}
`;

exports[`AttributeSimpleLinkRepo getReverseValues Should return values for advanced reverse link attribute into simple link 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "@value0": "test_lib",
    "@value2": "test_linked_lib",
    "value1": "123456",
    "value3": "test_simple_link_attr",
  },
  "query": "
                    FOR r IN @@value0
                        FILTER r._key == @value1
                        FOR l IN @@value2
                            FILTER r.@value3 == l._key
                            RETURN l
                ",
}
`;

exports[`AttributeSimpleLinkRepo getReverseValues Should return values for advanced reverse link attribute into simple link 2`] = `undefined`;

exports[`AttributeSimpleLinkRepo getValues Should return values for simple link attribute 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "@value0": "test_lib",
    "@value2": "test_linked_lib",
    "value1": "123456",
    "value3": "test_simple_link_attr",
  },
  "query": "
                    FOR r IN @@value0
                        FILTER r._key == @value1
                        FOR l IN @@value2
                            FILTER r.@value3 == l._key
                            RETURN l
                ",
}
`;

exports[`AttributeSimpleLinkRepo getValues Should return values for simple link attribute 2`] = `undefined`;

exports[`AttributeSimpleLinkRepo sortQueryPart Should return simple link sort 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": {},
    "value1": "label",
    "value2": "linked",
    "value3": "ASC",
  },
  "query": "SORT 
                FIRST(FOR l IN @value0
                    FILTER TO_STRING(r.@value1) == l._key
                RETURN l.@value2)
             @value3",
}
`;
